client-js/app/scripts/app.ts

```diff
@@ -319,49 +319,57 @@ window.addEventListener('WebComponentsReady', () => {
     const authMethods = await mgr.authMethods({manager: manager});
     app.host = manager;
     app.allowUsernameLogin = authMethods.indexOf('username+password') >= 0;
-    app.allowOAuthLogin = authMethods.indexOf('oauth_token') >= 0;
+    app.allowOAuthLogin = authMethods.indexOf('oauth_token') >= 0
+                       || authMethods.indexOf('oauth_access_token') >= 0;
     app.$.loginButton.addEventListener('click', (event) => doLogin());
     app.$.loginForm.addEventListener('submit', (event) => {
       event.preventDefault();
       event.stopPropagation();
       doLogin();
     });
     if (app.allowOAuthLogin) {
-      app.$.oauthButton.addEventListener('click', (event) => {
-        startOAuthLogin(mgr, manager)
+      app.$.oauthButton.addEventListener('click', async (event) => {
+        try {
+          const oauthInfo = await mgr.oauthInfo({manager: manager});
+          const tokenType = authMethods.indexOf('oauth_access_token') >= 0 ? 'access' : 'id';
+          const rememberMe = app.$.rememberPassword.checked;
+          await startOAuthLogin(manager, oauthInfo, tokenType, rememberMe)
+          // Redirected to oauth login page, so we never get here.
+        } catch (error) {
+          app.loginError = error.message || error;
+        }
       });
     }
     app.$.loginDialog.open();
     setTimeout(() => app.$.passwordInput.$.input.focus(), 0);
     return promise;
   }
 
-  async function startOAuthLogin(mgr: ManagerApi, manager: string) {
-    const rememberMe = app.$.rememberPassword.checked;
-    try {
-      const {clientId, clientSecret} = await mgr.oauthInfo({ manager: manager });
-      const redirectUri = `${location.protocol}//${location.host}${prefix}/oauth2callback`;
-      const params = {
-        'response_type': 'code',
-        'client_id': clientId,
-        'redirect_uri': redirectUri,
-        'scope': 'openid email profile',
-        'state': JSON.stringify({
-          path: location.pathname + location.search + location.hash,
-          manager: manager,
-          client_id: clientId,
-          client_secret: clientSecret,
-          redirect_uri: redirectUri,
-          remember_me: rememberMe
-        })
-      };
-      const baseUrl = 'https://accounts.google.com/o/oauth2/v2/auth'
-      const url = `${baseUrl}?${encodeQueryString(params)}`;
-      window.location.href = url;
-    } catch (error) {
-      app.loginError = error.message || error;
-      setTimeout(() => app.$.passwordInput.$.input.focus(), 0);
-    }
+  /**
+   * Start OAuth login flow by redirecting to oauth login page. Returns a
+   * promise that never fires since there's nothing to do after the redirect.
+   */
+  async function startOAuthLogin(manager: string, oauthInfo: OAuthInfo, tokenType: string, rememberMe: boolean) {
```

Comment:
Moved references to `app` into the login dialog event listeners above. Otherwise, this is mostly just outdenting the existing function.